src/quadtree.cpp

```diff
@@ -0,0 +1,199 @@
+
+	if (auto start_leaf = find_leaf_in_root(start_x_aligned, start_y_aligned)) {
+		Leaf* south_leaf = start_leaf;
+		Leaf* east_leaf;
+
+		for (int_fast32_t ny = start_y_aligned; ny <= end_y_aligned; ny += FLOOR_SIZE) {
+			east_leaf = south_leaf;
+
+			for (int_fast32_t nx = start_x_aligned; nx <= end_x_aligned; nx += FLOOR_SIZE) {
+				if (east_leaf) {
+					for (auto creature : east_leaf->creatures) {
+						if (comparasion(creature)) {
+							spectators.emplace_back(creature);
+						}
+					}
+
+					east_leaf = east_leaf->east_leaf;
+				} else {
+					east_leaf = find_leaf_in_root(nx + FLOOR_SIZE, ny);
+				}
+			}
+
+			if (south_leaf) {
+				south_leaf = south_leaf->south_leaf;
+			} else {
+				south_leaf = find_leaf_in_root(start_x_aligned, ny + FLOOR_SIZE);
+			}
+		}
+	}
+}
+
+Tile* tfs::quadtree::find_tile(uint16_t x, uint16_t y, uint8_t z)
+{
+	if (auto leaf = find_leaf_in_root(x, y)) {
+		return leaf->tiles[z][x & FLOOR_MASK][y & FLOOR_MASK];
+	}
+	return nullptr;
+}
+
+void tfs::quadtree::create_tile(uint16_t x, uint16_t y, uint8_t z, Tile* tile)
+{
+	create_leaf_in_root(x, y, z);
+
+	if (auto leaf = find_leaf_in_root(x, y)) {
+		leaf->tiles[z][x & FLOOR_MASK][y & FLOOR_MASK] = tile;
+	}
+}
+
+void tfs::quadtree::move_creature(uint16_t old_x, uint16_t old_y, uint8_t old_z, uint16_t x, uint16_t y, uint8_t z,
+                                  Creature* creature)
+{
+	if (auto old_leaf = find_leaf_in_root(old_x, old_y); auto leaf = find_leaf_in_root(x, y)) {
+		if (old_leaf != leaf) {
+			old_leaf->remove_creature(creature);
+			leaf->push_creature(creature);
+		}
+	}
+}
+
+void tfs::quadtree::insert_creature(uint16_t x, uint16_t y, uint8_t z, Creature* creature)
+{
+	if (auto leaf = find_leaf_in_root(x, y)) {
+		leaf->push_creature(creature);
+	}
+}
+
+void tfs::quadtree::remove_creature(uint16_t x, uint16_t y, uint8_t z, Creature* creature)
+{
+	if (auto leaf = find_leaf_in_root(x, y)) {
+		leaf->remove_creature(creature);
+	}
+}
+
+Leaf::Leaf(uint16_t x, uint16_t y)
+{
+	// update north
+	if (auto north_leaf = find_leaf_in_root(x, y - FLOOR_SIZE)) {
+		north_leaf->south_leaf = this;
+	}
+
+	// update west
+	if (auto west_leaf = find_leaf_in_root(x - FLOOR_SIZE, y)) {
+		west_leaf->east_leaf = this;
+	}
+
+	// update south
+	if (auto south_leaf = find_leaf_in_root(x, y + FLOOR_SIZE)) {
+		this->south_leaf = south_leaf;
+	}
+
+	// update east
+	if (auto east_leaf = find_leaf_in_root(x + FLOOR_SIZE, y)) {
+		this->east_leaf = east_leaf;
+	}
+}
+
+void Leaf::push_creature(Creature* creature)
+{
+	creatures.insert(creature);
+
+	if (auto monster = creature->getMonster()) {
+		monsters.insert(monster);
+	} else if (auto npc = creature->getNpc()) {
+		npcs.insert(npc);
+	} else if (auto player = creature->getPlayer()) {
+		players.insert(player);
+	}
+}
+
+void Leaf::remove_creature(Creature* creature)
+{
+	auto it = creatures.find(creature);
+	if (it != creatures.end()) {
+		creatures.erase(it);
+	}
```

Comment:
No need to check for `erase`, if the key is not present it does nothing

Reply:
No problem, I kept the search as it was a `vector`, but I changed it to a `set`. ;)